http3/request_writer.go

```diff
@@ -0,0 +1,312 @@
+package http3
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/lucas-clemente/quic-go/internal/utils"
+	"github.com/marten-seemann/qpack"
+	"golang.org/x/net/http/httpguts"
+	"golang.org/x/net/http2/hpack"
+	"golang.org/x/net/idna"
+)
+
+type requestWriter struct {
+	mutex     sync.Mutex
+	encoder   *qpack.Encoder
+	headerBuf *bytes.Buffer
+
+	logger utils.Logger
+}
+
+func newRequestWriter(logger utils.Logger) *requestWriter {
+	headerBuf := &bytes.Buffer{}
+	encoder := qpack.NewEncoder(headerBuf)
+	return &requestWriter{
+		encoder:   encoder,
+		headerBuf: headerBuf,
+		logger:    logger,
+	}
+}
+
+func (w *requestWriter) WriteRequest(str quic.Stream, req *http.Request) error {
+	headers, err := w.getHeaders(req)
+	if err != nil {
+		return err
+	}
+	if _, err := str.Write(headers); err != nil {
+		return err
+	}
+	// TODO: add support for trailers
+	if req.Body == nil {
+		str.Close()
+		return nil
+	}
+
+	// send the request body asynchronously
+	go func() {
+		if err := w.sendRequestBody(req.Body, str); err != nil {
+			w.logger.Errorf("Error writing request: %s", err)
+			return
+		}
+		str.Close()
+	}()
+
+	return nil
+}
+
+func (w *requestWriter) getHeaders(req *http.Request) ([]byte, error) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	defer w.encoder.Close()
+
+	if err := w.encodeHeaders(req, false, "", actualContentLength(req)); err != nil {
+		return nil, err
+	}
+
+	buf := &bytes.Buffer{}
+	hf := headersFrame{Length: uint64(w.headerBuf.Len())}
+	hf.Write(buf)
+	if _, err := io.Copy(buf, w.headerBuf); err != nil {
+		return nil, err
+	}
+	w.headerBuf.Reset()
+	return buf.Bytes(), nil
+}
+
+func (w *requestWriter) sendRequestBody(req io.ReadCloser, str quic.Stream) error {
+	b := make([]byte, 8*1024)
```

Comment:
Are the 8k taken from the spec?

Reply:
The spec doesn't give any guidance about this. Ideally, we'd have the request body as a `[]byte`, and just send a single DATA frame. However, `http.Request.Body` is a `io.ReadCloser`, so we need to copy the data and frame it. 8k is just a number I made up.